{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "50e8d179",
   "metadata": {},
   "source": [
    "# Binding prediction using AptaNet\n",
    "Step-by-step guide for using AptaNet for binary aptamerâ€“protein binding prediction."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a64683",
   "metadata": {},
   "source": [
    "## Overview\n",
    "This notebook showcases the AptaNet algorithm, a deep learning method that combines sequence-derived (k-mer + PSeAAC) features with RandomForest-based feature selection, and a multi-layer perceptron to predict whether an aptamer and a protein interact (binary classification: aptamer binds/does not bind with the target protein). An overview of the classes and helper functions used in this notebook:\n",
    "\n",
    "- **pairs_to_features**: helper that converts `(aptamer_sequence, protein_sequence)` pairs into feature vectors using k-mer + PSeAAC.\n",
    "- **SkorchAptaNet**: a PyTorch MLP wrapped in Skorch for binary classification.\n",
    "- **load_1gnh**: helper to load the 1GNH molecule structure from PDB file into our in-memory `MoleculeLoader` format."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45c592d9",
   "metadata": {},
   "source": [
    "## Data preparation\n",
    "To train the `AptaNetMLP` the notebook uses:\n",
    "* For `X`:\n",
    "    * 5 random aptamer sequences of length>30 (to satisfy the default lambda value of 30 set in the PSeAAC algorithm).\n",
    "    * Amino acid sequences extracted from the 1GNH protein molecule.\n",
    "    \n",
    "    The aptamer sequences and the amino acid sequences form tuples `(aptamer_sequence, protein_sequence)` to form `X`.\n",
    "* For `y`:\n",
    "    * A random binary value (to indicate if the aptamer binds or not) as dummy data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3737da88",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data imports\n",
    "import torch\n",
    "\n",
    "from pyaptamer.datasets import load_1gnh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a2f6701d",
   "metadata": {},
   "outputs": [],
   "source": [
    "aptamer_sequence = [\n",
    "    \"GGGAGGACGAAGACGACUCGAGACAGGCUAGGGAGGGA\",\n",
    "    \"AAGCGUCGGAUCUACACGUGCGAUAGCUCAGUACGCGGU\",\n",
    "    \"CGGUAUCGAGUACAGGAGUCCGACGGAUAGUCCGGAGC\",\n",
    "    \"UAGCUAGCGAACUAGGCGUAGCUUCGAGUAGCUACGGAA\",\n",
    "    \"GCUAGGACGAUCGCACGUGACCGUCAGUAGCGUAGGAGA\",\n",
    "]\n",
    "\n",
    "gnh = load_1gnh()\n",
    "protein_sequence = gnh.to_df_seq()[\"sequence\"].tolist()\n",
    "\n",
    "# Build all combinations (aptamer, protein), duplicated to increase dataset size\n",
    "X = [(a, p) for a in aptamer_sequence for p in protein_sequence] * 5\n",
    "\n",
    "# Dummy binary labels for the pairs\n",
    "y = torch.randint(0, 2, (len(X),), dtype=torch.float32)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c10ae4ea",
   "metadata": {},
   "source": [
    "## Building the pipeline\n",
    "### Dataset balancing using the Neighbourhood cleaning rule\n",
    "In the `AptaNet` paper, the authors mention using the `NeighbourhoodCleaningRule` from `imblearn` in order to balance the dataset, as in their dataset they had more negative (0) values than positives (1).\n",
    "\n",
    " To build a scikit-learn pipeline, follow these steps:\n",
    "1. Convert the input to the desired (aptamer_sequence, protein_sequence) format.\n",
    "    * OPTIONAL: As mentioned in the paper, perform under-sampling using the  \n",
    "    Neighborhood Cleaning Rule to balance the classes.\n",
    "2. Get the PSeAAC feature vectors for your converted input (using `pairs_to_features`).\n",
    "3. Select the number of features to use from the feature vector (using `RandomForestClassifier`).\n",
    "4. Define the skorch neural network (using `AptaNetMLP`).\n",
    "### Different workflows\n",
    "In this first half of the notebook we will cover 3 different workflows one can follow, in ascending order of customizability:\n",
    "\n",
    "1. A minimal workflow with no dataset balancing, while using the in-built pipeline.\n",
    "2. Using your own custom pipeline.\n",
    "3. Dataset balancing, while using your own pipeline."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b7c1997",
   "metadata": {},
   "source": [
    "### First workflow\n",
    "A minimal workflow with no dataset balancing, while using the in-built pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4609ce9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Suppress warnings for cleaner output\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "44cc1cbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you want to use the AptaNet pipeline, you can import it directly\n",
    "from pyaptamer.aptanet import AptaNetPipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d782bca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pipeline = AptaNetPipeline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f8a0d947",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fit the pipeline on the aptamer-protein pairs\n",
    "pipeline.fit(X, y)\n",
    "\n",
    "# Predict the labels for the training data\n",
    "y_pred = pipeline.predict(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ae896e5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training Accuracy: 0.64\n"
     ]
    }
   ],
   "source": [
    "# Optional: Evaluate training accuracy\n",
    "from sklearn.metrics import accuracy_score\n",
    "\n",
    "print(\"Training Accuracy:\", accuracy_score(y, y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b90a04f9",
   "metadata": {},
   "source": [
    "### Second Worflow\n",
    "\n",
    "Your own custom-built pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6130027a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you want to build your own aptamer pipeline, you should use the following imports\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.feature_selection import SelectFromModel\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.preprocessing import FunctionTransformer\n",
    "from skorch import NeuralNetBinaryClassifier\n",
    "\n",
    "from pyaptamer.aptanet._aptanet_nn import AptaNetMLP\n",
    "from pyaptamer.utils._aptanet_utils import pairs_to_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4a2e277e",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_transformer = FunctionTransformer(\n",
    "    func=pairs_to_features,\n",
    "    validate=False,\n",
    "    # Optional arguments for pairs_to_features\n",
    "    # example: kw_args={'k': 4, 'pseaac_kwargs': {'lambda_value': 30}}\n",
    "    kw_args={},\n",
    ")\n",
    "\n",
    "selector = SelectFromModel(\n",
    "    estimator=RandomForestClassifier(\n",
    "        n_estimators=300,\n",
    "        max_depth=9,\n",
    "        random_state=None,\n",
    "    ),\n",
    "    threshold=\"mean\",\n",
    ")\n",
    "\n",
    "# Define the classifier\n",
    "net = NeuralNetBinaryClassifier(\n",
    "    module=AptaNetMLP,\n",
    "    module__input_dim=None,\n",
    "    module__hidden_dim=128,\n",
    "    module__n_hidden=7,\n",
    "    module__dropout=0.3,\n",
    "    module__output_dim=1,\n",
    "    module__use_lazy=True,\n",
    "    criterion=torch.nn.BCEWithLogitsLoss,\n",
    "    max_epochs=200,\n",
    "    lr=0.00014,\n",
    "    optimizer=torch.optim.RMSprop,\n",
    "    optimizer__alpha=0.9,\n",
    "    optimizer__eps=1e-08,\n",
    "    device=\"cuda\" if torch.cuda.is_available() else \"cpu\",\n",
    ")\n",
    "\n",
    "pipeline = Pipeline(\n",
    "    [\n",
    "        (\"features\", feature_transformer),\n",
    "        (\"selector\", selector),\n",
    "        (\"clf\", net),\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8049a472",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  epoch    train_loss    valid_acc    valid_loss     dur\n",
      "-------  ------------  -----------  ------------  ------\n",
      "      1        \u001b[36m0.7558\u001b[0m       \u001b[32m0.2000\u001b[0m        \u001b[35m0.7114\u001b[0m  0.0068\n",
      "      2        \u001b[36m0.6823\u001b[0m       0.2000        \u001b[35m0.7110\u001b[0m  0.0078\n",
      "      3        0.7409       0.2000        \u001b[35m0.7105\u001b[0m  0.0083\n",
      "      4        0.7738       0.2000        \u001b[35m0.7101\u001b[0m  0.0074\n",
      "      5        0.7082       0.2000        \u001b[35m0.7098\u001b[0m  0.0070\n",
      "      6        \u001b[36m0.6472\u001b[0m       0.2000        \u001b[35m0.7095\u001b[0m  0.0068\n",
      "      7        0.6722       0.2000        \u001b[35m0.7092\u001b[0m  0.0079\n",
      "      8        0.7144       0.2000        \u001b[35m0.7089\u001b[0m  0.0056\n",
      "      9        0.6902       0.2000        \u001b[35m0.7087\u001b[0m  0.0072\n",
      "     10        0.6749       0.2000        \u001b[35m0.7085\u001b[0m  0.0082\n",
      "     11        0.6673       0.2000        \u001b[35m0.7083\u001b[0m  0.0065\n",
      "     12        0.7027       0.2000        \u001b[35m0.7081\u001b[0m  0.0073\n",
      "     13        0.7487       0.2000        \u001b[35m0.7079\u001b[0m  0.0081\n",
      "     14        0.6552       0.2000        \u001b[35m0.7077\u001b[0m  0.0063\n",
      "     15        0.6756       0.2000        \u001b[35m0.7075\u001b[0m  0.0080\n",
      "     16        0.6839       0.2000        \u001b[35m0.7073\u001b[0m  0.0062\n",
      "     17        0.7348       0.2000        \u001b[35m0.7071\u001b[0m  0.0074\n",
      "     18        0.7453       0.2000        \u001b[35m0.7069\u001b[0m  0.0080\n",
      "     19        0.8265       0.2000        \u001b[35m0.7067\u001b[0m  0.0076\n",
      "     20        0.7283       0.2000        \u001b[35m0.7064\u001b[0m  0.0086\n",
      "     21        0.7438       0.2000        \u001b[35m0.7064\u001b[0m  0.0084\n",
      "     22        0.6965       0.2000        \u001b[35m0.7061\u001b[0m  0.0076\n",
      "     23        0.7157       0.2000        \u001b[35m0.7060\u001b[0m  0.0082\n",
      "     24        0.7552       0.2000        \u001b[35m0.7058\u001b[0m  0.0066\n",
      "     25        0.7149       0.2000        \u001b[35m0.7057\u001b[0m  0.0089\n",
      "     26        0.6679       0.2000        \u001b[35m0.7056\u001b[0m  0.0069\n",
      "     27        0.7343       0.2000        \u001b[35m0.7054\u001b[0m  0.0072\n",
      "     28        0.6735       0.2000        \u001b[35m0.7053\u001b[0m  0.0087\n",
      "     29        0.7022       0.2000        \u001b[35m0.7050\u001b[0m  0.0083\n",
      "     30        0.6681       0.2000        \u001b[35m0.7048\u001b[0m  0.0068\n",
      "     31        0.7337       0.2000        \u001b[35m0.7046\u001b[0m  0.0083\n",
      "     32        0.7294       0.2000        \u001b[35m0.7044\u001b[0m  0.0084\n",
      "     33        0.7752       0.2000        \u001b[35m0.7042\u001b[0m  0.0067\n",
      "     34        0.6733       0.2000        \u001b[35m0.7040\u001b[0m  0.0082\n",
      "     35        0.7211       0.2000        \u001b[35m0.7039\u001b[0m  0.0082\n",
      "     36        0.6556       0.2000        \u001b[35m0.7036\u001b[0m  0.0087\n",
      "     37        0.6754       0.2000        \u001b[35m0.7035\u001b[0m  0.0086\n",
      "     38        0.7472       0.2000        \u001b[35m0.7033\u001b[0m  0.0087\n",
      "     39        0.7175       0.2000        \u001b[35m0.7031\u001b[0m  0.0095\n",
      "     40        0.7047       0.2000        \u001b[35m0.7031\u001b[0m  0.0078\n",
      "     41        0.7311       0.2000        \u001b[35m0.7030\u001b[0m  0.0082\n",
      "     42        0.7108       0.2000        \u001b[35m0.7027\u001b[0m  0.0092\n",
      "     43        0.6557       0.2000        \u001b[35m0.7025\u001b[0m  0.0070\n",
      "     44        0.7393       0.2000        \u001b[35m0.7023\u001b[0m  0.0086\n",
      "     45        0.7204       0.2000        \u001b[35m0.7021\u001b[0m  0.0076\n",
      "     46        0.6957       0.2000        \u001b[35m0.7019\u001b[0m  0.0081\n",
      "     47        0.8070       0.2000        \u001b[35m0.7018\u001b[0m  0.0086\n",
      "     48        0.8228       0.2000        \u001b[35m0.7016\u001b[0m  0.0094\n",
      "     49        0.6894       0.2000        \u001b[35m0.7015\u001b[0m  0.0062\n",
      "     50        \u001b[36m0.6270\u001b[0m       0.2000        \u001b[35m0.7012\u001b[0m  0.0076\n",
      "     51        0.6575       0.2000        \u001b[35m0.7011\u001b[0m  0.0088\n",
      "     52        0.7267       0.2000        \u001b[35m0.7009\u001b[0m  0.0064\n",
      "     53        0.7213       0.2000        \u001b[35m0.7007\u001b[0m  0.0090\n",
      "     54        0.7649       0.2000        \u001b[35m0.7005\u001b[0m  0.0089\n",
      "     55        0.6726       0.2000        \u001b[35m0.7005\u001b[0m  0.0085\n",
      "     56        0.7220       0.2000        \u001b[35m0.7003\u001b[0m  0.0080\n",
      "     57        0.7059       0.2000        \u001b[35m0.7001\u001b[0m  0.0082\n",
      "     58        0.7412       0.2000        \u001b[35m0.6999\u001b[0m  0.0064\n",
      "     59        0.7679       0.2000        \u001b[35m0.6998\u001b[0m  0.0067\n",
      "     60        0.7402       0.2000        \u001b[35m0.6997\u001b[0m  0.0078\n",
      "     61        0.7064       0.2000        \u001b[35m0.6996\u001b[0m  0.0080\n",
      "     62        0.7319       0.2000        \u001b[35m0.6993\u001b[0m  0.0071\n",
      "     63        0.7715       0.2000        \u001b[35m0.6991\u001b[0m  0.0064\n",
      "     64        0.6713       0.2000        \u001b[35m0.6988\u001b[0m  0.0075\n",
      "     65        0.7029       0.2000        \u001b[35m0.6987\u001b[0m  0.0077\n",
      "     66        0.8803       0.2000        \u001b[35m0.6986\u001b[0m  0.0094\n",
      "     67        0.7353       0.2000        \u001b[35m0.6982\u001b[0m  0.0076\n",
      "     68        0.7615       0.2000        \u001b[35m0.6981\u001b[0m  0.0082\n",
      "     69        0.7668       0.2000        \u001b[35m0.6979\u001b[0m  0.0078\n",
      "     70        0.7022       0.2000        \u001b[35m0.6977\u001b[0m  0.0080\n",
      "     71        0.6964       0.2000        \u001b[35m0.6975\u001b[0m  0.0081\n",
      "     72        0.7864       0.2000        \u001b[35m0.6973\u001b[0m  0.0083\n",
      "     73        0.6501       0.2000        \u001b[35m0.6971\u001b[0m  0.0072\n",
      "     74        0.7791       0.2000        \u001b[35m0.6971\u001b[0m  0.0081\n",
      "     75        0.6930       0.2000        \u001b[35m0.6970\u001b[0m  0.0078\n",
      "     76        0.7754       0.2000        \u001b[35m0.6968\u001b[0m  0.0080\n",
      "     77        0.7831       0.2000        \u001b[35m0.6967\u001b[0m  0.0075\n",
      "     78        0.6927       0.2000        \u001b[35m0.6965\u001b[0m  0.0096\n",
      "     79        0.7418       0.2000        \u001b[35m0.6963\u001b[0m  0.0084\n",
      "     80        0.7624       0.2000        \u001b[35m0.6961\u001b[0m  0.0070\n",
      "     81        0.7524       0.2000        \u001b[35m0.6959\u001b[0m  0.0081\n",
      "     82        0.6555       0.2000        \u001b[35m0.6957\u001b[0m  0.0069\n",
      "     83        0.7131       0.2000        \u001b[35m0.6956\u001b[0m  0.0077\n",
      "     84        0.7253       0.2000        \u001b[35m0.6954\u001b[0m  0.0080\n",
      "     85        0.6828       0.2000        \u001b[35m0.6953\u001b[0m  0.0069\n",
      "     86        0.7401       0.2000        \u001b[35m0.6950\u001b[0m  0.0070\n",
      "     87        0.7263       0.2000        \u001b[35m0.6949\u001b[0m  0.0083\n",
      "     88        0.7051       0.2000        \u001b[35m0.6945\u001b[0m  0.0088\n",
      "     89        0.7875       0.2000        \u001b[35m0.6944\u001b[0m  0.0083\n",
      "     90        0.8103       0.2000        \u001b[35m0.6942\u001b[0m  0.0085\n",
      "     91        0.6953       0.2000        \u001b[35m0.6939\u001b[0m  0.0086\n",
      "     92        0.6626       0.2000        \u001b[35m0.6937\u001b[0m  0.0081\n",
      "     93        0.6529       0.2000        \u001b[35m0.6936\u001b[0m  0.0077\n",
      "     94        0.7008       0.2000        \u001b[35m0.6935\u001b[0m  0.0082\n",
      "     95        0.8119       0.2000        \u001b[35m0.6933\u001b[0m  0.0087\n",
      "     96        0.6419       0.2000        \u001b[35m0.6932\u001b[0m  0.0070\n",
      "     97        0.6761       0.2000        \u001b[35m0.6932\u001b[0m  0.0095\n",
      "     98        0.7203       \u001b[32m0.8000\u001b[0m        \u001b[35m0.6930\u001b[0m  0.0091\n",
      "     99        0.6917       0.8000        \u001b[35m0.6928\u001b[0m  0.0107\n",
      "    100        0.7093       0.8000        \u001b[35m0.6925\u001b[0m  0.0081\n",
      "    101        0.7234       0.8000        \u001b[35m0.6924\u001b[0m  0.0128\n",
      "    102        0.6822       0.8000        \u001b[35m0.6922\u001b[0m  0.0094\n",
      "    103        0.6473       0.8000        \u001b[35m0.6921\u001b[0m  0.0110\n",
      "    104        0.6987       0.8000        \u001b[35m0.6919\u001b[0m  0.0096\n",
      "    105        0.7724       0.8000        \u001b[35m0.6918\u001b[0m  0.0115\n",
      "    106        0.7206       0.8000        \u001b[35m0.6918\u001b[0m  0.0090\n",
      "    107        0.7448       0.8000        \u001b[35m0.6916\u001b[0m  0.0091\n",
      "    108        0.7574       0.8000        \u001b[35m0.6915\u001b[0m  0.0091\n",
      "    109        0.6687       0.8000        \u001b[35m0.6912\u001b[0m  0.0091\n",
      "    110        0.6649       0.8000        \u001b[35m0.6911\u001b[0m  0.0089\n",
      "    111        0.8026       0.8000        \u001b[35m0.6910\u001b[0m  0.0094\n",
      "    112        0.6905       0.8000        \u001b[35m0.6909\u001b[0m  0.0091\n",
      "    113        0.7166       0.8000        \u001b[35m0.6908\u001b[0m  0.0062\n",
      "    114        0.7362       0.8000        \u001b[35m0.6907\u001b[0m  0.0096\n",
      "    115        0.7028       0.8000        \u001b[35m0.6904\u001b[0m  0.0094\n",
      "    116        0.7705       0.8000        \u001b[35m0.6903\u001b[0m  0.0090\n",
      "    117        0.7541       0.8000        \u001b[35m0.6903\u001b[0m  0.0080\n",
      "    118        0.6642       0.8000        \u001b[35m0.6901\u001b[0m  0.0081\n",
      "    119        0.6437       0.8000        \u001b[35m0.6899\u001b[0m  0.0075\n",
      "    120        0.6942       0.8000        \u001b[35m0.6898\u001b[0m  0.0095\n",
      "    121        0.6394       0.8000        \u001b[35m0.6896\u001b[0m  0.0078\n",
      "    122        0.7981       0.8000        \u001b[35m0.6894\u001b[0m  0.0081\n",
      "    123        0.7899       0.8000        \u001b[35m0.6891\u001b[0m  0.0088\n",
      "    124        0.7084       0.8000        0.6892  0.0073\n",
      "    125        0.6876       0.8000        \u001b[35m0.6889\u001b[0m  0.0095\n",
      "    126        0.6425       0.8000        \u001b[35m0.6887\u001b[0m  0.0091\n",
      "    127        0.6982       0.8000        \u001b[35m0.6886\u001b[0m  0.0076\n",
      "    128        \u001b[36m0.6229\u001b[0m       0.8000        \u001b[35m0.6885\u001b[0m  0.0088\n",
      "    129        0.6773       0.8000        \u001b[35m0.6883\u001b[0m  0.0083\n",
      "    130        0.6754       0.8000        \u001b[35m0.6880\u001b[0m  0.0074\n",
      "    131        0.7976       0.8000        \u001b[35m0.6878\u001b[0m  0.0119\n",
      "    132        0.7689       0.8000        \u001b[35m0.6877\u001b[0m  0.0119\n",
      "    133        0.7429       0.8000        \u001b[35m0.6876\u001b[0m  0.0113\n",
      "    134        0.6415       0.8000        \u001b[35m0.6874\u001b[0m  0.0111\n",
      "    135        0.7838       0.8000        \u001b[35m0.6873\u001b[0m  0.0097\n",
      "    136        0.7471       0.8000        \u001b[35m0.6871\u001b[0m  0.0084\n",
      "    137        0.7074       0.8000        \u001b[35m0.6869\u001b[0m  0.0072\n",
      "    138        0.8197       0.8000        \u001b[35m0.6867\u001b[0m  0.0082\n",
      "    139        0.7870       0.8000        \u001b[35m0.6865\u001b[0m  0.0083\n",
      "    140        0.7599       0.8000        \u001b[35m0.6863\u001b[0m  0.0074\n",
      "    141        0.6776       0.8000        \u001b[35m0.6862\u001b[0m  0.0096\n",
      "    142        0.6951       0.8000        \u001b[35m0.6861\u001b[0m  0.0099\n",
      "    143        0.7641       0.8000        \u001b[35m0.6860\u001b[0m  0.0085\n",
      "    144        0.7112       0.8000        \u001b[35m0.6858\u001b[0m  0.0086\n",
      "    145        0.6850       0.8000        \u001b[35m0.6856\u001b[0m  0.0084\n",
      "    146        \u001b[36m0.6151\u001b[0m       0.8000        \u001b[35m0.6855\u001b[0m  0.0086\n",
      "    147        0.7002       0.8000        \u001b[35m0.6854\u001b[0m  0.0088\n",
      "    148        0.7573       0.8000        \u001b[35m0.6852\u001b[0m  0.0073\n",
      "    149        0.6366       0.8000        \u001b[35m0.6850\u001b[0m  0.0092\n",
      "    150        0.6738       0.8000        \u001b[35m0.6848\u001b[0m  0.0083\n",
      "    151        0.8104       0.8000        \u001b[35m0.6847\u001b[0m  0.0086\n",
      "    152        0.6580       0.8000        \u001b[35m0.6845\u001b[0m  0.0064\n",
      "    153        0.7475       0.8000        \u001b[35m0.6843\u001b[0m  0.0114\n",
      "    154        0.7590       0.8000        \u001b[35m0.6842\u001b[0m  0.0103\n",
      "    155        0.8110       0.8000        \u001b[35m0.6841\u001b[0m  0.0090\n",
      "    156        0.7156       0.8000        \u001b[35m0.6839\u001b[0m  0.0072\n",
      "    157        0.6793       0.8000        \u001b[35m0.6836\u001b[0m  0.0074\n",
      "    158        0.7928       0.8000        \u001b[35m0.6835\u001b[0m  0.0080\n",
      "    159        0.8416       0.8000        \u001b[35m0.6833\u001b[0m  0.0069\n",
      "    160        0.7104       0.8000        \u001b[35m0.6831\u001b[0m  0.0092\n",
      "    161        0.7289       0.8000        \u001b[35m0.6830\u001b[0m  0.0097\n",
      "    162        0.7229       0.8000        \u001b[35m0.6829\u001b[0m  0.0108\n",
      "    163        0.7034       0.8000        \u001b[35m0.6827\u001b[0m  0.0546\n",
      "    164        0.7646       0.8000        \u001b[35m0.6825\u001b[0m  0.0132\n",
      "    165        0.7541       0.8000        \u001b[35m0.6823\u001b[0m  0.0088\n",
      "    166        0.6248       0.8000        \u001b[35m0.6821\u001b[0m  0.0085\n",
      "    167        0.6975       0.8000        0.6822  0.0090\n",
      "    168        0.7189       0.8000        \u001b[35m0.6821\u001b[0m  0.0093\n",
      "    169        0.6629       0.8000        \u001b[35m0.6820\u001b[0m  0.0094\n",
      "    170        0.6663       0.8000        \u001b[35m0.6816\u001b[0m  0.0087\n",
      "    171        0.7668       0.8000        \u001b[35m0.6815\u001b[0m  0.0082\n",
      "    172        0.7025       0.8000        \u001b[35m0.6812\u001b[0m  0.0084\n",
      "    173        0.7603       0.8000        \u001b[35m0.6811\u001b[0m  0.0086\n",
      "    174        0.6245       0.8000        \u001b[35m0.6810\u001b[0m  0.0075\n",
      "    175        0.6585       0.8000        \u001b[35m0.6809\u001b[0m  0.0083\n",
      "    176        0.6789       0.8000        \u001b[35m0.6808\u001b[0m  0.0085\n",
      "    177        0.7748       0.8000        \u001b[35m0.6807\u001b[0m  0.0070\n",
      "    178        0.7191       0.8000        \u001b[35m0.6805\u001b[0m  0.0081\n",
      "    179        0.7285       0.8000        \u001b[35m0.6802\u001b[0m  0.0102\n",
      "    180        0.7054       0.8000        \u001b[35m0.6801\u001b[0m  0.0070\n",
      "    181        0.7214       0.8000        \u001b[35m0.6800\u001b[0m  0.0083\n",
      "    182        0.7077       0.8000        \u001b[35m0.6798\u001b[0m  0.0093\n",
      "    183        0.7264       0.8000        \u001b[35m0.6796\u001b[0m  0.0091\n",
      "    184        0.8109       0.8000        \u001b[35m0.6795\u001b[0m  0.0081\n",
      "    185        0.6886       0.8000        \u001b[35m0.6793\u001b[0m  0.0086\n",
      "    186        0.7855       0.8000        \u001b[35m0.6791\u001b[0m  0.0089\n",
      "    187        0.8580       0.8000        \u001b[35m0.6791\u001b[0m  0.0084\n",
      "    188        0.7328       0.8000        \u001b[35m0.6791\u001b[0m  0.0074\n",
      "    189        0.7229       0.8000        \u001b[35m0.6789\u001b[0m  0.0119\n",
      "    190        0.6328       0.8000        \u001b[35m0.6787\u001b[0m  0.0092\n",
      "    191        0.6653       0.8000        \u001b[35m0.6786\u001b[0m  0.0073\n",
      "    192        0.7991       0.8000        \u001b[35m0.6785\u001b[0m  0.0084\n",
      "    193        0.7363       0.8000        \u001b[35m0.6784\u001b[0m  0.0087\n",
      "    194        0.8113       0.8000        \u001b[35m0.6782\u001b[0m  0.0067\n",
      "    195        0.7330       0.8000        \u001b[35m0.6780\u001b[0m  0.0086\n",
      "    196        0.7514       0.8000        \u001b[35m0.6778\u001b[0m  0.0083\n",
      "    197        0.6795       0.8000        \u001b[35m0.6777\u001b[0m  0.0063\n",
      "    198        0.6814       0.8000        \u001b[35m0.6776\u001b[0m  0.0071\n",
      "    199        0.8200       0.8000        \u001b[35m0.6774\u001b[0m  0.0080\n",
      "    200        0.6446       0.8000        \u001b[35m0.6772\u001b[0m  0.0077\n"
     ]
    }
   ],
   "source": [
    "# Fit the pipeline on the aptamer-protein pairs\n",
    "pipeline.fit(X, y)\n",
    "\n",
    "# Predict the labels for the training data\n",
    "y_pred = pipeline.predict(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f00091db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training Accuracy: 0.64\n"
     ]
    }
   ],
   "source": [
    "# Optional: Evaluate training accuracy\n",
    "from sklearn.metrics import accuracy_score\n",
    "\n",
    "print(\"Training Accuracy:\", accuracy_score(y, y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37db693e",
   "metadata": {},
   "source": [
    "### Third workflow\n",
    "Dataset balancing using under-sampling, while using your own pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "523f4ce6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you want to build your own aptamer pipeline, you should use the following imports\n",
    "from sklearn.preprocessing import FunctionTransformer\n",
    "\n",
    "from pyaptamer.aptanet import AptaNetClassifier, AptaNetPipeline\n",
    "from pyaptamer.utils._aptanet_utils import pairs_to_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0cf0bed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTIONAL: If you want to use the Neighborhood Cleaning Rule for under-sampling\n",
    "# NOTE: If you want to use under-sampling, you need to install imbalanced-learn\n",
    "# and use the Pipeline from imblearn\n",
    "# %pip install imblearn\n",
    "from imblearn.pipeline import Pipeline\n",
    "from imblearn.under_sampling import NeighbourhoodCleaningRule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "13fb1669",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_transformer = FunctionTransformer(\n",
    "    func=pairs_to_features,\n",
    "    validate=False,\n",
    "    # Optional arguments for pairs_to_features\n",
    "    # example: kw_args={'k': 4, 'pseaac_kwargs': {'lambda_value': 30}}\n",
    "    kw_args={},\n",
    ")\n",
    "\n",
    "# AptaNetClassifier encapsulates the \"selector\" and \"net\" mentioned in Workflow 2\n",
    "clf = AptaNetClassifier()\n",
    "\n",
    "model = Pipeline(\n",
    "    [\n",
    "        (\"ncr\", NeighbourhoodCleaningRule()),\n",
    "        (\"clf\", clf),\n",
    "    ]\n",
    ")\n",
    "\n",
    "pipeline = AptaNetPipeline(estimator=model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "1ed76399",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fit the pipeline on the aptamer-protein pairs\n",
    "pipeline.fit(X, y)\n",
    "\n",
    "# Predict the labels for the training data\n",
    "y_pred = pipeline.predict(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "945dc876",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training Accuracy: 0.36\n"
     ]
    }
   ],
   "source": [
    "# Optional: Evaluate training accuracy\n",
    "from sklearn.metrics import accuracy_score\n",
    "\n",
    "print(\"Training Accuracy:\", accuracy_score(y, y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0947e5d2",
   "metadata": {},
   "source": [
    "# Implementing `AptaNet` for real-world use cases"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94894a1b",
   "metadata": {},
   "source": [
    "## Data preparation\n",
    "To train the `AptaNetMLP` the notebook uses the dataset used to train the `AptaTrans` algorithm, this dataset can be found in `pyaptamer/datasets/data/train_li2014`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "273e2dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data imports\n",
    "\n",
    "from pyaptamer.datasets import load_li2014"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d5cb05fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "X, y = load_li2014(split=\"train\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "5f1442f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aptamer</th>\n",
       "      <th>protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UCGGAGGUGGUUCAGCUCUGCAUCGACAGUUGGC</td>\n",
       "      <td>MDSTNYISKLFEYAQRQGQISDIKFEEVGTDGPDHLKTFTLRVVIK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CGGGGTGTTGTCCTGTGCTCTCCGGAGAGCACAGGACAACACCCCG</td>\n",
       "      <td>MDELFPLIFPAEPAQASGPYVEIIEQPKQRGMRFRYKCEGRSAGSI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTGCAGTTGCACTGAATTCGCCTCTCGCCTCCGTACACTTAGTCG...</td>\n",
       "      <td>MDEVGAQVAAPMFIHQSLGRKRDLYYPMSNRLVQSQPQRRDEWNSK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TATTTGCCCTTGCAGGCCGCAGGAGTGCTAGCAGT</td>\n",
       "      <td>PISPIDTVPVKLKPGMDGPKVKQWPLTEEKIKALTEICNEMEKEGK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UCUCUGGGCUCUUAGGAGAACGGAUAGGAGUGUGCUCGCU</td>\n",
       "      <td>LQENFLPQPRQQHHGTLVLHYRPHREEAGMQHPCLWPGSSHCSDDS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             aptamer  \\\n",
       "0                 UCGGAGGUGGUUCAGCUCUGCAUCGACAGUUGGC   \n",
       "1     CGGGGTGTTGTCCTGTGCTCTCCGGAGAGCACAGGACAACACCCCG   \n",
       "2  GCTGCAGTTGCACTGAATTCGCCTCTCGCCTCCGTACACTTAGTCG...   \n",
       "3                TATTTGCCCTTGCAGGCCGCAGGAGTGCTAGCAGT   \n",
       "4           UCUCUGGGCUCUUAGGAGAACGGAUAGGAGUGUGCUCGCU   \n",
       "\n",
       "                                             protein  \n",
       "0  MDSTNYISKLFEYAQRQGQISDIKFEEVGTDGPDHLKTFTLRVVIK...  \n",
       "1  MDELFPLIFPAEPAQASGPYVEIIEQPKQRGMRFRYKCEGRSAGSI...  \n",
       "2  MDEVGAQVAAPMFIHQSLGRKRDLYYPMSNRLVQSQPQRRDEWNSK...  \n",
       "3  PISPIDTVPVKLKPGMDGPKVKQWPLTEEKIKALTEICNEMEKEGK...  \n",
       "4  LQENFLPQPRQQHHGTLVLHYRPHREEAGMQHPCLWPGSSHCSDDS...  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display first few rows of X which contains aptamer-protein pairs\n",
    "X.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2a63d1db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      label\n",
       "0  positive\n",
       "1  positive\n",
       "2  positive\n",
       "3  positive\n",
       "4  positive"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display first few rows of y which contains binary binding labels\n",
    "y.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01001dd2",
   "metadata": {},
   "source": [
    "## DIfferent Real-world examples\n",
    "In the second half of this notebook we will explore 3 real-world use cases of the algorithm. These include:\n",
    "1. Training on the entire AptaTrans dataset, and predict binding probability (`predict_proba`) between a protein and a DNA sequence.\n",
    "2. TODO: Same for using DNA sequences from a `fasta` file.\n",
    "3. Using MCTS combined with a trained `AptaNet` to propose new aptamers for a new pdb file, i.e., a form of in-silico Selex."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "71f1648e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fit the pipeline on the aptamer-protein pairs from the `AptaTrans` dataset\n",
    "pipeline = AptaNetPipeline()\n",
    "pipeline.fit(X, y)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94949073",
   "metadata": {},
   "source": [
    "### First workflow\n",
    "\n",
    "Predicting binding probability (`predict_proba`) between a protein (1GNH) and a DNA sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "7c5a42a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.9648417 , 0.03515826]], dtype=float32)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aptamer_sequence = \"GGGAGGACGAAGACGACUCGAGACAGGCUAGGGAGGGA\"\n",
    "\n",
    "X = [(aptamer_sequence, p) for p in protein_sequence]\n",
    "\n",
    "pipeline.predict_proba(X)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af501b0c",
   "metadata": {},
   "source": [
    "### Second workflow (TODO)\n",
    "\n",
    "Predicting binding probability (`predict_proba`) between a protein (1GNH) and DNA sequences from a `fasta` file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16420bc4",
   "metadata": {},
   "source": [
    "### Third workflow\n",
    "\n",
    "Using MCTS combined with a trained `AptaNet` to propose new aptamers for a protein (1GNH), i.e., a form of in-silico Selex."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "aadff246",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ----- Round: 1 -----\n",
      "##################################################\n",
      "Best subsequence: _CC_U_\n",
      "Depth: 3\n",
      "##################################################\n",
      "\n",
      " ----- Round: 2 -----\n",
      "##################################################\n",
      "Best subsequence: _CC_U_C__GC_\n",
      "Depth: 6\n",
      "##################################################\n",
      "\n",
      " ----- Round: 3 -----\n",
      "##################################################\n",
      "Best subsequence: _CC_U_C__GC__CA__C\n",
      "Depth: 9\n",
      "##################################################\n",
      "\n",
      " ----- Round: 4 -----\n",
      "##################################################\n",
      "Best subsequence: _CC_U_C__GC__CA__CC_C_U_\n",
      "Depth: 12\n",
      "##################################################\n",
      "\n",
      " ----- Round: 5 -----\n",
      "##################################################\n",
      "Best subsequence: _CC_U_C__GC__CA__CC_C_U_C__U_U\n",
      "Depth: 15\n",
      "##################################################\n",
      "\n",
      " ----- Round: 6 -----\n",
      "##################################################\n",
      "Best subsequence: _CC_U_C__GC__CA__CC_C_U_C__U_UC_A_U_\n",
      "Depth: 18\n",
      "##################################################\n",
      "\n",
      " ----- Round: 7 -----\n",
      "##################################################\n",
      "Best subsequence: _CC_U_C__GC__CA__CC_C_U_C__U_UC_A_U__CG_\n",
      "Depth: 20\n",
      "##################################################\n"
     ]
    }
   ],
   "source": [
    "from pyaptamer.experiments import AptamerEvalAptaNet\n",
    "from pyaptamer.mcts import MCTS\n",
    "\n",
    "# There can only be one target sequence\n",
    "eval = AptamerEvalAptaNet(pipeline=pipeline, target=protein_sequence[0])\n",
    "\n",
    "mcts = MCTS(experiment=eval)\n",
    "\n",
    "output = mcts.run(verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "d4397f07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GUACCUCCACCUCCGCCUUC'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the best sequence\n",
    "output[\"candidate\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c19b2b6b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyaptamer-latest",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
