{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "50e8d179",
   "metadata": {},
   "source": [
    "# Binding prediction using AptaNet\n",
    "Step-by-step guide for using AptaNet for binary aptamerâ€“protein binding prediction."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a64683",
   "metadata": {},
   "source": [
    "## Overview\n",
    "This notebook showcases the AptaNet algorithm, a deep learning method that combines sequence-derived (k-mer + PSeAAC) features with RandomForest-based feature selection, and a multi-layer perceptron to predict whether an aptamer and a protein interact (binary classification: aptamer binds/does not bind with the target protein). An overview of the classes and helper functions used in this notebook:\n",
    "\n",
    "- **pairs_to_features**: helper that converts `(aptamer_sequence, protein_sequence)` pairs into feature vectors using k-mer + PSeAAC.\n",
    "- **SkorchAptaNet**: a PyTorch MLP wrapped in Skorch for binary classification.\n",
    "- **load_1gnh_structure**: helper to load the 1GNH molecule structure from PDB file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45c592d9",
   "metadata": {},
   "source": [
    "## Data preparation\n",
    "To train the `AptaNetMLP` the notebook uses:\n",
    "* For `X`:\n",
    "    * 5 random aptamer sequences of length>30 (to satisfy the default lambda value of 30 set in the PSeAAC algorithm).\n",
    "    * Amino acid sequences extracted from the 1GNH protein molecule.\n",
    "    \n",
    "    The aptamer sequences and the amino acid sequences form tuples `(aptamer_sequence, protein_sequence)` to form `X`.\n",
    "* For `y`:\n",
    "    * A random binary value (to indicate if the aptamer binds or not) as dummy data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3737da88",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data imports\n",
    "import torch\n",
    "\n",
    "from pyaptamer.datasets import load_1gnh_structure\n",
    "from pyaptamer.utils import struct_to_aaseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a2f6701d",
   "metadata": {},
   "outputs": [],
   "source": [
    "aptamer_sequence = [\n",
    "    \"GGGAGGACGAAGACGACUCGAGACAGGCUAGGGAGGGA\",\n",
    "    \"AAGCGUCGGAUCUACACGUGCGAUAGCUCAGUACGCGGU\",\n",
    "    \"CGGUAUCGAGUACAGGAGUCCGACGGAUAGUCCGGAGC\",\n",
    "    \"UAGCUAGCGAACUAGGCGUAGCUUCGAGUAGCUACGGAA\",\n",
    "    \"GCUAGGACGAUCGCACGUGACCGUCAGUAGCGUAGGAGA\",\n",
    "]\n",
    "\n",
    "gnh = load_1gnh_structure()\n",
    "protein_sequence = struct_to_aaseq(gnh)\n",
    "\n",
    "# Build all combinations (aptamer, protein)\n",
    "X = [(a, p) for a in aptamer_sequence for p in protein_sequence]\n",
    "\n",
    "# Dummy binary labels for the pairs\n",
    "y = torch.randint(0, 2, (len(X),), dtype=torch.float32)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c10ae4ea",
   "metadata": {},
   "source": [
    "## Building the pipeline\n",
    "### Dataset balancing using the Neighbourhood cleaning rule\n",
    "In the `AptaNet` paper, the authors mention using the `NeighbourhoodCleaningRule` from `imblearn` in order to balance the dataset, as in their dataset they had more negative (0) values than positives (1).\n",
    "\n",
    " To build a scikit-learn pipeline, follow these steps:\n",
    "1. Convert the input to the desired (aptamer_sequence, protein_sequence) format.\n",
    "    * OPTIONAL: As mentioned in the paper, perform under-sampling using the  \n",
    "    Neighborhood Cleaning Rule to balance the classes.\n",
    "2. Get the PSeAAC feature vectors for your converted input (using `pairs_to_features`).\n",
    "3. Select the number of features to use from the feature vector (using `RandomForestClassifier`).\n",
    "4. Define the skorch neural network (using `AptaNetMLP`).\n",
    "### Different workflows\n",
    "In this first half of the notebook we will cover 3 different workflows one can follow, in ascending order of customizability:\n",
    "\n",
    "1. A minimal workflow with no dataset balancing, while using the in-built pipeline.\n",
    "2. Using your own custom pipeline.\n",
    "3. Dataset balancing, while using your own pipeline."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b7c1997",
   "metadata": {},
   "source": [
    "### First workflow\n",
    "A minimal workflow with no dataset balancing, while using the in-built pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "44cc1cbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you want to use the AptaNet pipeline, you can import it directly\n",
    "from pyaptamer.aptanet import AptaNetPipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d782bca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pipeline = AptaNetPipeline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f8a0d947",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fit the pipeline on the aptamer-protein pairs\n",
    "pipeline.fit(X, y)\n",
    "\n",
    "# Predict the labels for the training data\n",
    "y_pred = pipeline.predict(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ae896e5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training Accuracy: 0.48\n"
     ]
    }
   ],
   "source": [
    "# Optional: Evaluate training accuracy\n",
    "from sklearn.metrics import accuracy_score\n",
    "\n",
    "print(\"Training Accuracy:\", accuracy_score(y, y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b90a04f9",
   "metadata": {},
   "source": [
    "### Second Worflow\n",
    "\n",
    "Your own custom-built pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6130027a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you want to build your own aptamer pipeline, you should use the following imports\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.feature_selection import SelectFromModel\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.preprocessing import FunctionTransformer\n",
    "from skorch import NeuralNetBinaryClassifier\n",
    "\n",
    "from pyaptamer.aptanet._aptanet_nn import AptaNetMLP\n",
    "from pyaptamer.utils._aptanet_utils import pairs_to_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4a2e277e",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_transformer = FunctionTransformer(\n",
    "    func=pairs_to_features,\n",
    "    validate=False,\n",
    "    # Optional arguments for pairs_to_features\n",
    "    # example: kw_args={'k': 4, 'pseaac_kwargs': {'lambda_value': 30}}\n",
    "    kw_args={},\n",
    ")\n",
    "\n",
    "selector = SelectFromModel(\n",
    "    estimator=RandomForestClassifier(\n",
    "        n_estimators=300,\n",
    "        max_depth=9,\n",
    "        random_state=None,\n",
    "    ),\n",
    "    threshold=\"mean\",\n",
    ")\n",
    "\n",
    "# Define the classifier\n",
    "net = NeuralNetBinaryClassifier(\n",
    "    module=AptaNetMLP,\n",
    "    module__input_dim=None,\n",
    "    module__hidden_dim=128,\n",
    "    module__n_hidden=7,\n",
    "    module__dropout=0.3,\n",
    "    module__output_dim=1,\n",
    "    module__use_lazy=True,\n",
    "    criterion=torch.nn.BCEWithLogitsLoss,\n",
    "    max_epochs=200,\n",
    "    lr=0.00014,\n",
    "    optimizer=torch.optim.RMSprop,\n",
    "    optimizer__alpha=0.9,\n",
    "    optimizer__eps=1e-08,\n",
    "    device=\"cuda\" if torch.cuda.is_available() else \"cpu\",\n",
    ")\n",
    "\n",
    "pipeline = Pipeline(\n",
    "    [\n",
    "        (\"features\", feature_transformer),\n",
    "        (\"selector\", selector),\n",
    "        (\"clf\", net),\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "8049a472",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  epoch    train_loss    valid_acc    valid_loss     dur\n",
      "-------  ------------  -----------  ------------  ------\n",
      "      1        \u001b[36m0.6816\u001b[0m       \u001b[32m0.5000\u001b[0m        \u001b[35m0.6935\u001b[0m  0.0081\n",
      "      2        0.7010       0.5000        0.6935  0.0090\n",
      "      3        0.7804       0.5000        0.6935  0.0090\n",
      "      4        0.7011       0.5000        0.6935  0.0110\n",
      "      5        \u001b[36m0.6776\u001b[0m       0.5000        0.6936  0.0110\n",
      "      6        0.7262       0.5000        0.6936  0.0090\n",
      "      7        0.6914       0.5000        0.6936  0.0100\n",
      "      8        0.7281       0.5000        0.6936  0.0100\n",
      "      9        0.7713       0.5000        0.6936  0.0125\n",
      "     10        0.7409       0.5000        0.6936  0.0110\n",
      "     11        \u001b[36m0.6732\u001b[0m       0.5000        0.6936  0.0150\n",
      "     12        0.7088       0.5000        0.6936  0.0090\n",
      "     13        0.8154       0.5000        0.6936  0.0089\n",
      "     14        0.6870       0.5000        0.6936  0.0090\n",
      "     15        0.6901       0.5000        0.6936  0.0090\n",
      "     16        0.7002       0.5000        0.6936  0.0090\n",
      "     17        0.7640       0.5000        0.6936  0.0105\n",
      "     18        0.7396       0.5000        0.6936  0.0090\n",
      "     19        \u001b[36m0.6671\u001b[0m       0.5000        0.6936  0.0080\n",
      "     20        0.7470       0.5000        0.6936  0.0100\n",
      "     21        0.7711       0.5000        0.6936  0.0100\n",
      "     22        0.7018       0.5000        0.6936  0.0089\n",
      "     23        0.7094       0.5000        0.6936  0.0100\n",
      "     24        0.6954       0.5000        0.6936  0.0111\n",
      "     25        0.7355       0.5000        0.6936  0.0124\n",
      "     26        0.6843       0.5000        0.6936  0.0090\n",
      "     27        0.7300       0.5000        0.6936  0.0080\n",
      "     28        0.7056       0.5000        0.6936  0.0130\n",
      "     29        0.7262       0.5000        0.6936  0.0120\n",
      "     30        0.7494       0.5000        0.6936  0.0100\n",
      "     31        0.7414       0.5000        0.6936  0.0110\n",
      "     32        0.7667       0.5000        0.6936  0.0135\n",
      "     33        0.7079       0.5000        0.6936  0.0090\n",
      "     34        0.7174       0.5000        0.6936  0.0130\n",
      "     35        0.7529       0.5000        0.6936  0.0090\n",
      "     36        0.7628       0.5000        0.6936  0.0087\n",
      "     37        0.7058       0.5000        0.6936  0.0119\n",
      "     38        0.7381       0.5000        0.6936  0.0080\n",
      "     39        0.7490       0.5000        0.6936  0.0070\n",
      "     40        \u001b[36m0.6117\u001b[0m       0.5000        0.6936  0.0080\n",
      "     41        0.7694       0.5000        0.6936  0.0064\n",
      "     42        0.7059       0.5000        0.6936  0.0070\n",
      "     43        0.7059       0.5000        0.6936  0.0080\n",
      "     44        0.7729       0.5000        0.6936  0.0070\n",
      "     45        0.7679       0.5000        0.6936  0.0080\n",
      "     46        0.6798       0.5000        0.6936  0.0070\n",
      "     47        0.7279       0.5000        0.6937  0.0100\n",
      "     48        0.7527       0.5000        0.6937  0.0070\n",
      "     49        0.7342       0.5000        0.6937  0.0070\n",
      "     50        0.7332       0.5000        0.6937  0.0080\n",
      "     51        0.7274       0.5000        0.6937  0.0065\n",
      "     52        0.7959       0.5000        0.6937  0.0060\n",
      "     53        0.7840       0.5000        0.6937  0.0070\n",
      "     54        0.7578       0.5000        0.6937  0.0090\n",
      "     55        0.6882       0.5000        0.6937  0.0100\n",
      "     56        0.6825       0.5000        0.6937  0.0080\n",
      "     57        0.7623       0.5000        0.6937  0.0070\n",
      "     58        0.6957       0.5000        0.6937  0.0080\n",
      "     59        0.7085       0.5000        0.6937  0.0070\n",
      "     60        0.6646       0.5000        0.6937  0.0070\n",
      "     61        0.7268       0.5000        0.6937  0.0070\n",
      "     62        0.6805       0.5000        0.6937  0.0120\n",
      "     63        0.7425       0.5000        0.6937  0.0080\n",
      "     64        0.7610       0.5000        0.6937  0.0090\n",
      "     65        0.6750       0.5000        0.6937  0.0070\n",
      "     66        0.6685       0.5000        0.6937  0.0090\n",
      "     67        0.6342       0.5000        0.6937  0.0080\n",
      "     68        0.6578       0.5000        0.6937  0.0060\n",
      "     69        0.7222       0.5000        0.6937  0.0080\n",
      "     70        0.7625       0.5000        0.6937  0.0060\n",
      "     71        0.6975       0.5000        0.6937  0.0070\n",
      "     72        0.7292       0.5000        0.6937  0.0065\n",
      "     73        0.6989       0.5000        0.6937  0.0060\n",
      "     74        0.7089       0.5000        0.6937  0.0060\n",
      "     75        0.8102       0.5000        0.6937  0.0060\n",
      "     76        0.6702       0.5000        0.6937  0.0060\n",
      "     77        0.7568       0.5000        0.6937  0.0060\n",
      "     78        0.7512       0.5000        0.6937  0.0060\n",
      "     79        0.7106       0.5000        0.6937  0.0060\n",
      "     80        0.7104       0.5000        0.6937  0.0060\n",
      "     81        0.6601       0.5000        0.6937  0.0060\n",
      "     82        0.7910       0.5000        0.6937  0.0050\n",
      "     83        0.7923       0.5000        0.6937  0.0070\n",
      "     84        0.6951       0.5000        0.6937  0.0060\n",
      "     85        0.6744       0.5000        0.6937  0.0062\n",
      "     86        0.7410       0.5000        0.6937  0.0060\n",
      "     87        0.7445       0.5000        0.6937  0.0060\n",
      "     88        0.7394       0.5000        0.6937  0.0060\n",
      "     89        0.6974       0.5000        0.6937  0.0060\n",
      "     90        0.7438       0.5000        0.6937  0.0060\n",
      "     91        0.7256       0.5000        0.6937  0.0065\n",
      "     92        0.7731       0.5000        0.6937  0.0060\n",
      "     93        0.7100       0.5000        0.6937  0.0060\n",
      "     94        0.7273       0.5000        0.6937  0.0060\n",
      "     95        0.6877       0.5000        0.6937  0.0060\n",
      "     96        0.7127       0.5000        0.6937  0.0090\n",
      "     97        0.7528       0.5000        0.6937  0.0055\n",
      "     98        0.7548       0.5000        0.6937  0.0060\n",
      "     99        0.7031       0.5000        0.6937  0.0060\n",
      "    100        0.6954       0.5000        0.6937  0.0080\n",
      "    101        0.7099       0.5000        0.6937  0.0060\n",
      "    102        0.6733       0.5000        0.6937  0.0060\n",
      "    103        0.6626       0.5000        0.6937  0.0080\n",
      "    104        0.7132       0.5000        0.6937  0.0080\n",
      "    105        0.6578       0.5000        0.6937  0.0090\n",
      "    106        0.7144       0.5000        0.6937  0.0080\n",
      "    107        0.7589       0.5000        0.6937  0.0060\n",
      "    108        0.6908       0.5000        0.6937  0.0060\n",
      "    109        0.7501       0.5000        0.6937  0.0063\n",
      "    110        0.6649       0.5000        0.6937  0.0090\n",
      "    111        0.6968       0.5000        0.6937  0.0090\n",
      "    112        0.6888       0.5000        0.6937  0.0070\n",
      "    113        0.7275       0.5000        0.6937  0.0070\n",
      "    114        0.7153       0.5000        0.6937  0.0080\n",
      "    115        0.7026       0.5000        0.6937  0.0080\n",
      "    116        0.7802       0.5000        0.6937  0.0060\n",
      "    117        0.7314       0.5000        0.6937  0.0060\n",
      "    118        0.7274       0.5000        0.6937  0.0070\n",
      "    119        0.6905       0.5000        0.6937  0.0070\n",
      "    120        0.7010       0.5000        0.6937  0.0080\n",
      "    121        0.6984       0.5000        0.6937  0.0070\n",
      "    122        0.6451       0.5000        0.6937  0.0080\n",
      "    123        0.6709       0.5000        0.6937  0.0080\n",
      "    124        0.7027       0.5000        0.6938  0.0060\n",
      "    125        0.8258       0.5000        0.6938  0.0100\n",
      "    126        0.7098       0.5000        0.6938  0.0110\n",
      "    127        0.7317       0.5000        0.6938  0.0090\n",
      "    128        0.6790       0.5000        0.6938  0.0100\n",
      "    129        0.8289       0.5000        0.6938  0.0075\n",
      "    130        0.6928       0.5000        0.6938  0.0060\n",
      "    131        0.7276       0.5000        0.6938  0.0070\n",
      "    132        0.7704       0.5000        0.6938  0.0070\n",
      "    133        0.7191       0.5000        0.6938  0.0080\n",
      "    134        0.7134       0.5000        0.6938  0.0090\n",
      "    135        0.6912       0.5000        0.6938  0.0070\n",
      "    136        0.6965       0.5000        0.6938  0.0090\n",
      "    137        0.6787       0.5000        0.6938  0.0060\n",
      "    138        0.7227       0.5000        0.6938  0.0060\n",
      "    139        0.7283       0.5000        0.6938  0.0080\n",
      "    140        \u001b[36m0.5687\u001b[0m       0.5000        0.6938  0.0075\n",
      "    141        0.6830       0.5000        0.6938  0.0070\n",
      "    142        0.7023       0.5000        0.6938  0.0070\n",
      "    143        0.6991       0.5000        0.6938  0.0070\n",
      "    144        0.6996       0.5000        0.6938  0.0090\n",
      "    145        0.7420       0.5000        0.6938  0.0065\n",
      "    146        0.7173       0.5000        0.6938  0.0080\n",
      "    147        0.6840       0.5000        0.6938  0.0090\n",
      "    148        0.7580       0.5000        0.6938  0.0070\n",
      "    149        0.7435       0.5000        0.6938  0.0080\n",
      "    150        0.7484       0.5000        0.6938  0.0065\n",
      "    151        0.6979       0.5000        0.6938  0.0080\n",
      "    152        0.7080       0.5000        0.6938  0.0090\n",
      "    153        0.7272       0.5000        0.6938  0.0090\n",
      "    154        0.7120       0.5000        0.6938  0.0090\n",
      "    155        0.7852       0.5000        0.6938  0.0070\n",
      "    156        0.7241       0.5000        0.6938  0.0060\n",
      "    157        0.7240       0.5000        0.6938  0.0070\n",
      "    158        0.6459       0.5000        0.6938  0.0060\n",
      "    159        0.6417       0.5000        0.6938  0.0080\n",
      "    160        0.7578       0.5000        0.6938  0.0070\n",
      "    161        0.7706       0.5000        0.6938  0.0075\n",
      "    162        0.6543       0.5000        0.6938  0.0060\n",
      "    163        0.6663       0.5000        0.6938  0.0070\n",
      "    164        0.7494       0.5000        0.6938  0.0070\n",
      "    165        0.7175       0.5000        0.6938  0.0080\n",
      "    166        0.6841       0.5000        0.6938  0.0070\n",
      "    167        0.7393       0.5000        0.6938  0.0070\n",
      "    168        0.7187       0.5000        0.6938  0.0060\n",
      "    169        0.6869       0.5000        0.6938  0.0070\n",
      "    170        0.7566       0.5000        0.6938  0.0090\n",
      "    171        0.7465       0.5000        0.6938  0.0070\n",
      "    172        0.7675       0.5000        0.6938  0.0080\n",
      "    173        0.7077       0.5000        0.6938  0.0080\n",
      "    174        0.7220       0.5000        0.6938  0.0070\n",
      "    175        0.7054       0.5000        0.6938  0.0060\n",
      "    176        0.7202       0.5000        0.6938  0.0070\n",
      "    177        0.7247       0.5000        0.6938  0.0070\n",
      "    178        0.7142       0.5000        0.6938  0.0100\n",
      "    179        0.7057       0.5000        0.6938  0.0060\n",
      "    180        0.7602       0.5000        0.6938  0.0080\n",
      "    181        0.7917       0.5000        0.6938  0.0070\n",
      "    182        0.7675       0.5000        0.6938  0.0090\n",
      "    183        0.6947       0.5000        0.6938  0.0090\n",
      "    184        0.7238       0.5000        0.6938  0.0070\n",
      "    185        0.6834       0.5000        0.6938  0.0080\n",
      "    186        0.7217       0.5000        0.6938  0.0070\n",
      "    187        0.6933       0.5000        0.6938  0.0083\n",
      "    188        0.7040       0.5000        0.6938  0.0100\n",
      "    189        0.7235       0.5000        0.6938  0.0070\n",
      "    190        0.7306       0.5000        0.6938  0.0070\n",
      "    191        0.7542       0.5000        0.6938  0.0090\n",
      "    192        0.7507       0.5000        0.6938  0.0095\n",
      "    193        0.7325       0.5000        0.6938  0.0070\n",
      "    194        0.6789       0.5000        0.6938  0.0070\n",
      "    195        0.7666       0.5000        0.6938  0.0080\n",
      "    196        0.7337       0.5000        0.6938  0.0110\n",
      "    197        0.7269       0.5000        0.6938  0.0075\n",
      "    198        0.7311       0.5000        0.6938  0.0070\n",
      "    199        0.8000       0.5000        0.6938  0.0070\n",
      "    200        0.7251       0.5000        0.6938  0.0080\n"
     ]
    }
   ],
   "source": [
    "# Fit the pipeline on the aptamer-protein pairs\n",
    "pipeline.fit(X, y)\n",
    "\n",
    "# Predict the labels for the training data\n",
    "y_pred = pipeline.predict(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f00091db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training Accuracy: 0.52\n"
     ]
    }
   ],
   "source": [
    "# Optional: Evaluate training accuracy\n",
    "from sklearn.metrics import accuracy_score\n",
    "\n",
    "print(\"Training Accuracy:\", accuracy_score(y, y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37db693e",
   "metadata": {},
   "source": [
    "### Third workflow\n",
    "Dataset balancing using under-sampling, while using your own pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "523f4ce6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you want to build your own aptamer pipeline, you should use the following imports\n",
    "from sklearn.preprocessing import FunctionTransformer\n",
    "\n",
    "from pyaptamer.aptanet import AptaNetClassifier, AptaNetPipeline\n",
    "from pyaptamer.utils._aptanet_utils import pairs_to_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0cf0bed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTIONAL: If you want to use the Neighborhood Cleaning Rule for under-sampling\n",
    "# NOTE: If you want to use under-sampling, you need to install imbalanced-learn\n",
    "# and use the Pipeline from imblearn\n",
    "# %pip install imblearn\n",
    "from imblearn.pipeline import Pipeline\n",
    "from imblearn.under_sampling import NeighbourhoodCleaningRule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "13fb1669",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_transformer = FunctionTransformer(\n",
    "    func=pairs_to_features,\n",
    "    validate=False,\n",
    "    # Optional arguments for pairs_to_features\n",
    "    # example: kw_args={'k': 4, 'pseaac_kwargs': {'lambda_value': 30}}\n",
    "    kw_args={},\n",
    ")\n",
    "\n",
    "# AptaNetClassifier encapsulates the \"selector\" and \"net\" mentioned in Workflow 2\n",
    "clf = AptaNetClassifier()\n",
    "\n",
    "model = Pipeline(\n",
    "    [\n",
    "        (\"ncr\", NeighbourhoodCleaningRule()),\n",
    "        (\"clf\", clf),\n",
    "    ]\n",
    ")\n",
    "\n",
    "pipeline = AptaNetPipeline(estimator=model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1ed76399",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\satvm\\miniconda3\\envs\\pyaptamer-latest\\Lib\\site-packages\\sklearn\\model_selection\\_split.py:811: UserWarning: The least populated class in y has only 2 members, which is less than n_splits=5.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Fit the pipeline on the aptamer-protein pairs\n",
    "pipeline.fit(X, y)\n",
    "\n",
    "# Predict the labels for the training data\n",
    "y_pred = pipeline.predict(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "945dc876",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training Accuracy: 0.48\n"
     ]
    }
   ],
   "source": [
    "# Optional: Evaluate training accuracy\n",
    "from sklearn.metrics import accuracy_score\n",
    "\n",
    "print(\"Training Accuracy:\", accuracy_score(y, y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0947e5d2",
   "metadata": {},
   "source": [
    "# Implementing `AptaNet` for real-world use cases"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94894a1b",
   "metadata": {},
   "source": [
    "## Data preparation\n",
    "To train the `AptaNetMLP` the notebook uses the dataset used to train the `AptaTrans` algorithm, this dataset can be found in `pyaptamer/datasets/data/train_li2014`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "273e2dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from pyaptamer.datasets import load_csv_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "d5cb05fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = load_csv_dataset(\"train_li2014\")\n",
    "\n",
    "df = pd.concat([df[:30], df[2300:]])\n",
    "\n",
    "X_raw, y_raw = df.iloc[:, :2], df.iloc[:, 2]\n",
    "\n",
    "# Build combinations in the form of (aptamer, protein)\n",
    "X = list(zip(X_raw.iloc[:, 0], X_raw.iloc[:, 1], strict=False))\n",
    "\n",
    "# Dummy binary labels for the pairs\n",
    "y = np.where(y_raw == \"positive\", 1, 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "9194427c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0]\n"
     ]
    }
   ],
   "source": [
    "print(y)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01001dd2",
   "metadata": {},
   "source": [
    "## DIfferent Real-world examples\n",
    "In the second half of this notebook we will explore 3 real-world use cases of the algorithm. These include:\n",
    "1. Training on the entire AptaTrans dataset, and predict binding probability (`predict_proba`) between a protein and a DNA sequence.\n",
    "2. same for using DNA sequences from a `fasta` file.\n",
    "3. Using MCTS combined with a trained `AptaNet` to propose new aptamers for a new pdb file, i.e., a form of in-silico Selex."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "71f1648e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\satvm\\pyaptamer\\pyaptamer\\pseaac\\_pseaac_aptanet.py:196: UserWarning: Invalid amino acid(s) found in sequence. Replaced with 'N'.\n",
      "  seq = clean_protein_seq(protein_sequence)\n"
     ]
    }
   ],
   "source": [
    "# Fit the pipeline on the aptamer-protein pairs from the `AptaTrans` dataset\n",
    "pipeline.fit(X, y)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94949073",
   "metadata": {},
   "source": [
    "### First workflow\n",
    "\n",
    "Predicting binding probability (`predict_proba`) between a protein (1GNH) and a DNA sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "7c5a42a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.53823733, 0.46176267],\n",
       "       [0.53823733, 0.46176267],\n",
       "       [0.53823733, 0.46176267],\n",
       "       [0.53823733, 0.46176267],\n",
       "       [0.53823733, 0.46176267],\n",
       "       [0.53823733, 0.46176267],\n",
       "       [0.53823733, 0.46176267],\n",
       "       [0.53823733, 0.46176267],\n",
       "       [0.53823733, 0.46176267],\n",
       "       [0.53823733, 0.46176267]], dtype=float32)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aptamer_sequence = \"GGGAGGACGAAGACGACUCGAGACAGGCUAGGGAGGGA\"\n",
    "\n",
    "gnh = load_1gnh_structure()\n",
    "protein_sequence = struct_to_aaseq(gnh)\n",
    "\n",
    "X = [(aptamer_sequence, p) for p in protein_sequence]\n",
    "\n",
    "pipeline.predict_proba(X)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af501b0c",
   "metadata": {},
   "source": [
    "### Second workflow\n",
    "\n",
    "Predicting binding probability (`predict_proba`) between a protein (1GNH) and DNA sequences from a `fasta` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "dc580069",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.53823286, 0.46176714],\n",
       "       [0.5382338 , 0.46176618],\n",
       "       [0.5382333 , 0.46176675],\n",
       "       [0.53823274, 0.46176726],\n",
       "       [0.53823316, 0.46176684],\n",
       "       [0.5382338 , 0.46176618],\n",
       "       [0.5382329 , 0.46176705],\n",
       "       [0.5382345 , 0.46176553],\n",
       "       [0.5382335 , 0.46176648],\n",
       "       [0.538234  , 0.46176603]], dtype=float32)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pyaptamer.trafos.convert import AnyToAASeq\n",
    "from pyaptamer.utils import hf_to_dataset\n",
    "\n",
    "ds = hf_to_dataset(\n",
    "    \"https://huggingface.co/datasets/gcos/HoloRBP4_round8_trimmed/resolve/main/HoloRBP4_round8_trimmed.fasta\",\n",
    "    download_locally=True,\n",
    ")\n",
    "\n",
    "transformer = AnyToAASeq(format=\"fasta\")\n",
    "fasta = transformer.transform(\"hf_datasets/HoloRBP4_round8_trimmed.fasta\")\n",
    "\n",
    "seqs = fasta[\"sequence\"].tolist()\n",
    "X = [(aptamer_sequence, seq) for seq in seqs]\n",
    "\n",
    "pipeline.predict_proba(X[:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16420bc4",
   "metadata": {},
   "source": [
    "### Third workflow\n",
    "\n",
    "Using MCTS combined with a trained `AptaNet` to propose new aptamers for a protein (1GNH), i.e., a form of in-silico Selex."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "aadff246",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ----- Round: 1 -----\n",
      "##################################################\n",
      "Best subsequence: _G_GG_\n",
      "Depth: 3\n",
      "##################################################\n",
      "\n",
      " ----- Round: 2 -----\n",
      "##################################################\n",
      "Best subsequence: _G_GG__A_GG_\n",
      "Depth: 6\n",
      "##################################################\n",
      "\n",
      " ----- Round: 3 -----\n",
      "##################################################\n",
      "Best subsequence: _G_GG__A_GG__U_A_A\n",
      "Depth: 9\n",
      "##################################################\n",
      "\n",
      " ----- Round: 4 -----\n",
      "##################################################\n",
      "Best subsequence: _G_GG__A_GG__U_A_AC__A_U\n",
      "Depth: 12\n",
      "##################################################\n",
      "\n",
      " ----- Round: 5 -----\n",
      "##################################################\n",
      "Best subsequence: _G_GG__A_GG__U_A_AC__A_U_AG__C\n",
      "Depth: 15\n",
      "##################################################\n",
      "\n",
      " ----- Round: 6 -----\n",
      "##################################################\n",
      "Best subsequence: _G_GG__A_GG__U_A_AC__A_U_AG__CC_G_U_\n",
      "Depth: 18\n",
      "##################################################\n",
      "\n",
      " ----- Round: 7 -----\n",
      "##################################################\n",
      "Best subsequence: _G_GG__A_GG__U_A_AC__A_U_AG__CC_G_U__A_G\n",
      "Depth: 20\n",
      "##################################################\n"
     ]
    }
   ],
   "source": [
    "from pyaptamer.experiments import AptamerEvalAptaNet\n",
    "from pyaptamer.mcts import MCTS\n",
    "\n",
    "gnh = load_1gnh_structure()\n",
    "protein_sequence = struct_to_aaseq(gnh)\n",
    "\n",
    "# There can only be one target sequence\n",
    "eval = AptamerEvalAptaNet(pipeline=pipeline, target=protein_sequence[0])\n",
    "\n",
    "mcts = MCTS(experiment=eval)\n",
    "\n",
    "output = mcts.run(verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "d4397f07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'UGCGCGGGGAGUAAAUACAG'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the best sequence\n",
    "output[\"candidate\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c19b2b6b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyaptamer-latest",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
